kf_lib_data_ingest/validation/data_validator.py

```diff
@@ -0,0 +1,352 @@
+                logger.info("Uninserted edges not stable yet.")
+                prev_edges = edges
+
+        def shake(edges):
+            # This is a bit like one iteration of a bidirectional bubble sort
+            edges = acbc_abc(edges)
+            edges = acab_abc(edges)
+            return edges
+
+        edges = prune_unneeded(try_until_stable(shake, edges))
+
+    # ###### Throw remaining edges into the graph ###### #
+    # ###### (I'm not actually sure if we should) ###### #
+    for e in edges:
+        graph.add_edge(a, b)
+
+    logger.info(f"Final edge count: {_get_edge_counts(graph)}")
+    return graph
+
+
+def _get_node_counts(graph):
+    """Count the number of unique instances of each type of value.
+
+    :param graph: a graph representation of the data
+    :return: dict of type keys and count values
+    """
+    return {c: len(graph.nodes(to_node=c)) for c in reversed(HIERARCHY.nodes())}
+
+
+def _get_edge_counts(graph):
+    hnodes = set(HIERARCHY.nodes())
+    return len([e for e in graph.edges() if e[1] not in hnodes])
+
+
+def _validate_graph_relationships(graph, dict_of_dataframes):
+    """Perform tests on the graph to validate which of the hierarchy
+    rules have been broken by the study data.
+
+    :param graph: a graph representation of the data
+    :param dict_of_dataframes: dict with filename keys and dataframe values
+    :yield: test result dicts
+    """
+    assert isinstance(graph, Graph)
+    assert len(graph.nodes()) > 0
+
+    logger.info("Validating graph relationships...")
+
+    membership_lookup = {
+        f: {col: set(df[col].values) for col in df.columns}
+        for f, df in dict_of_dataframes.items()
+    }
+
+    def find_in_files(node):
+        """Find which files a given node came from.
+
+        :param node: a node in the graph representing a data value
+        :return: a list of files
+        """
+        return [
+            f
+            for f, ml in membership_lookup.items()
+            if (node[0] in ml) and (node[1] in ml[node[0]])
+        ]
+
+    def cardinality(typeA, typeB, relation):
+        """Tests cardinality of connections between typeA and typeB.
+
+        :param typeA: a type node (nodes are inserted in the graph for value
+            types as well as values)
+        :param typeB: another type node
+        :param relation: one of the tuples from relations.RELATIONS
+        :return: a test result dict
+        """
+        desc, func = relation
+        logger.info(f"Testing {typeA} links to {desc} {typeB}...")
+        errors = []
+        A_nodes = graph.nodes(to_node=typeA)
+        use_from = typeB in HIERARCHY.nodes(typeA)
+        for n in A_nodes:
+            from_n = n if use_from else None
+            to_n = None if use_from else n
+            links = [c for c in graph.nodes(from_n, to_n) if c[0] == typeB]
+            if not func(len(links)):
+                locs = {m: find_in_files(m) for m in ([n] + links)}
```

Comment:
Iterating over all of the files to find a node seems like it should be done once at the beginning of validation or you could even store the files discovered with each node as you build the graph and then look up a node's files when you need them

Reply:
I considered putting an lru cache on it, but didn't think it would be sane to store a million copies of each file path instead of peeking at a few indexes. The membership_lookup structure generated at the beginning is pretty efficient.

Reply:
For clarity, I threw an lru_cache on find_in_files to see how often it would benefit, and on a test study with

```
DataValidator - Thread: MainThread - INFO - 1658 nodes added.
	SEQUENCING|INSTRUMENT               :
	SEQUENCING|PLATFORM                 : 2
	SEQUENCING|PAIRED_END               : 1
	SEQUENCING|STRATEGY                 : 3
	SEQUENCING|LIBRARY_NAME             :
	GENOMIC_FILE|REFERENCE_GENOME       : 1
	GENOMIC_FILE|HARMONIZED             : 1
	BIOSPECIMEN|SAMPLE_PROCUREMENT      :
	BIOSPECIMEN|VOLUME_UL               :
	BIOSPECIMEN|CONCENTRATION_MG_PER_ML :
	BIOSPECIMEN|SHIPMENT_DATE           :
	BIOSPECIMEN|SHIPMENT_ORIGIN         :
	BIOSPECIMEN|SPATIAL_DESCRIPTOR      :
	BIOSPECIMEN|EVENT_AGE_DAYS          : 97
	BIOSPECIMEN|TUMOR_DESCRIPTOR        : 4
	BIOSPECIMEN|ANATOMY_SITE            : 13
	BIOSPECIMEN|TISSUE_TYPE             : 2
	BIOSPECIMEN|COMPOSITION             : 2
	BIOSPECIMEN|ANALYTE                 : 1
	PARTICIPANT|ENROLLMENT_AGE_DAYS     :
	PARTICIPANT|SPECIES                 :
	PARTICIPANT|RACE                    :
	PARTICIPANT|ETHNICITY               :
	PARTICIPANT|GENDER                  : 2
	GENOMIC_FILE|ID                     : 442
	SEQUENCING|ID                       : 240
	BIOSPECIMEN|ID                      : 410
	BIOSPECIMEN_GROUP|ID                : 240
	FAMILY|ID                           :
	PARTICIPANT|ID                      : 197
```

The hit/miss numbers were 
```
CacheInfo(hits=32, misses=542, maxsize=1000, currsize=542)
```
So it got called 542 times, and only 32 calls were repeats.

Reply:
Oh I see, I'm sorry I didn't see that you generated that `membership_lookup` once at the beginning of validation.